docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,350 @@
+            uint32_t         : 16;   // reserved bits
+        };
+    } pfec;
+    uint32_t reserved; 
+} sgx_pfinfo;
+
+/**
+ * Custom page fault (#PF) handler, do usage specific processing upon #PF, e.g. loading data
+ * and verify its trustworthiness, then call sgx_mm_commit_data to explicitly EACCEPTCOPY data.
+ * This custom handler is passed into sgx_mm_alloc, and associated with the newly allocated region.
+ * The memory manager calls the handler when a #PF happens in the associated region.
+ * 
+ * @param[in] pfinfo, info reported in the SSA MISC region for page fault
+ * @param[in] private_data, private data provided by handler in sgx_mm_alloc call.
+ * @return EXCEPTION_CONTINUE_EXECUTION on success handling the exception.
+ *    abort if internal corruption detected. Otherwise EXCEPTION_CONTINUE_SEARCH.
+ */
+typedef int (*enclave_fault_handler_t)(const sgx_pfinfo* pfinfo, void* private_data);
+
+/**
+* Allocate a new memory region in enclave address space (ELRANGE).
+* @param[in] addr requested address, page aligned, if NULL is provided, then the function will select the address
+* @param[in] length size of the region in multiples of page size in bytes
+* @param[in] flags a bitwise OR of flags describing committing mode, committing order, address preference.
+*        Flags should include exactly one of following for committing mode:
+*            - EMA_RESERVE: just reserve an address range, no EPC commited. To allocate memory on a reserved range, 
+*                           call this function again with EMA_COMMIT_ON_DEMAND or EMA_COMMIT_NOW.
+*            - EMA_COMMIT_NOW: reserves memory range and commit EPC pages. EACCEPT for all allocated pages 
+*                              are done when this function returns.
+*            - EMA_COMMIT_ON_DEMAND: reserves memory range, EPC pages are committed (EACCEPT) on demand upon #PF.
+*        ORed with zero or one of the committing order flags:
+*            - EMA_GROWSDOWN: always commit pages from higher to lower addresses, no gaps in addresses above the last committed.
+*            - EMA_GROWSUP:   always commit pages from lower to higher addresses, no gaps in addresses below the last committed.
+*        Optionally ORed with EMA_FIXED to indicate allocation at fixed address. 
+*
+* @param[in] handler, custom handler for page faults in this region, NULL if no custom handling needed
+* @param[in] handler_private, private data for the @handler, which will be passed back when the handler is called.
+* @param[out] out_addr, pointer to store the start address of allocated range. Set to valid address by the function on success, NULL otherwise. 
+* @return 0 on success. EEXIST if range requested is overlapping a region in use and EMA_FIXED is requested,
+* ENOMEM for out of memory, EACCES if region is outside enclave address space.
+*/
+int sgx_mm_alloc(void *addr, size_t length, int flags, enclave_fault_handler_t handler, void* handler_private, void** out_addr)
```

Comment:
```suggestion
int sgx_mm_alloc(void *address, size_t length, int flags, enclave_fault_handler_t handler, void* handler_private, void** out_address)
```
OE style guidelines say to avoid abbreviations

Reply:
As we will position this as separate library, I'll keep abbreviations for now :-)